controller/handler_ctrl/bind.go

```diff
@@ -99,3 +119,60 @@ func (self *ctrlChannelLatencyHandler) HandleLatency(latencyType latency.Type, e
 		self.queueTimeHistogram.Update(elapsed.Nanoseconds())
 	}
 }
+
+type heartbeatCallback struct {
+	latencyMetric    metrics.Histogram
+	queueTimeMetric  metrics.Histogram
+	firstSent        int64
+	lastResponse     int64
+	ch               channel.Channel
+	latencySemaphore concurrenz.Semaphore
+}
+
+func (self *heartbeatCallback) HeartbeatTx(int64) {
+	if self.firstSent == 0 {
+		self.firstSent = time.Now().UnixMilli()
+	}
+}
+
+func (self *heartbeatCallback) HeartbeatRx(int64) {}
+
+func (self *heartbeatCallback) HeartbeatRespTx(int64) {}
+
+func (self *heartbeatCallback) HeartbeatRespRx(ts int64) {
+	now := time.Now()
+	self.lastResponse = now.UnixMilli()
```

Comment:
This should be `now.UnixNano()` since we've already grabbed the time just above. I put in a fix for the link version when I saw this.

Reply:
Moved to nano